Update erl_tar to support PAX format (redux)

bitwalker opened the pull request:
This commit introduces the following key changes:

- Support for reading tar archives in formats currently in common use,
such as v7, STAR, USTAR, PAX, and GNU tar's extensions to the
STAR/USTAR format.
- Support for writing PAX archives, only when necessary, using USTAR
when possible for greater portability.

These changes result in lifting of some prior restrictions:

- Support for reading archives produced by modern tar implementations
when other restrictions described below are present.
- Support for filenames which exceed 100 bytes in length, or paths which
exceed 255 bytes (see USTAR format specification for more details on
this restriction).
- Support for filenames of arbitrary length
- Support for unicode metadata (the previous behaviour of erl_tar was
actually violating the spec, by writing unicode-encoded data to fields
which are defined to be 7-bit ASCII, even though this technically
worked when using erl_tar at source and destination, it may not have
worked with other tar utilities, and this implementation now conforms
to the spec).
- Support for uid/gid values which cannot be converted to octal
integers.

I'm opening this PR to get discussion started around the changes. This changeset is significant, so I imagine it will take some time to review (I would suggest reading the source file over the diff, as it's likely to be more useful due to the size of the diff), and I'm expecting that I may need to change things to conform to your requirements. I've taken the following steps to ensure this PR is a good starting point:

- I've run the full OTP test suite, with no failures (at least that are caused by this PR, there were failures in unrelated places for what appear to be environmental reasons)
- I've added additional tests to cover new functionality
- I've additionally tested this implementation on archives created by both GNU tar, the old erl_tar, and BSD tar (libarchive). I also tested it with archives of different formats produced by GNU/BSD tar utilities, to better exercise it.
- I've updated the docs as necessary, though from an API point of view, nothing has changed
- I've added typespecs and run dialyzer on the new code, with no warnings or errors
- I've tried to stay true to the code style I've seen, and did my best to keep the formatting within 80 characters per line, though there are a couple of exceptions where they are a few characters more, but the readability gets trashed when forcing it under). If it's an absolute must, I can go back and hit those remaining spots.

Some of the internal changes I've made which I'm sure will spawn discussion:

- Created records rather than using raw tuples in some places, I've found this to be a big help in reading the code, as well as reasoning about it. While building this out, the use of records helped me catch numerous bugs in my initial implementation, and will continue to help prevent issues from being introduced in the future in my opinion.
- Moved all record and type definitions to erl_tar.hrl under lib/stdlib/src, to keep the erl_tar module itself more focused on the implementation code.

I want to get other eyes on this so that if there is anything missing, incorrect, redundant, etc., we can start catching it now. I'm very interested in whether anyone has better ideas on how to approach generating the header data as well - the implementation currently creates a ramfile initialized with an empty 512-byte block, and operates on that, as I could find no clean way to build up a binary while still handling the conditional code which changes what data is written to it. It unfortunately is very stateful, but is actually easier to reason about as a result. I'm certainly open to suggestions though.

See #1316 and #1268 for background which motivated this change.

bitwalker commented:
As per discussion in the last PR, I've addressed the dialyzer failures, rebased my commit on master, and added some more tests. Coverage is only at 69% currently, so I'll be adding more tests, but I wanted to get this back up so you all can get eyes on it.

bjorng commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -225,335 +278,1056 @@ month(10) -> "Oct";
+      }.
+
+to_sparse_array(Bin) when is_binary(Bin) ->
+    MaxEntries = trunc(byte_size(Bin) / 24),
```
Use integer division:

    MaxEntries = byte_size(Bin) div 24,

bitwalker replied:
Ah good call, no idea why I did it that way

bjorng commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -225,335 +278,1056 @@ month(10) -> "Oct";
+pad_file(Reader = #reader{pos=Pos}) ->
+    %% There must be at least two zero blocks at the end.
+    PadCurrent = skip_padding(Pos+?BLOCK_SIZE),
+    Padding = lists:duplicate(PadCurrent, 0),
```
Using the bit syntax should be faster:

    <<0:PadCurrent/unit:8>>

bitwalker replied:
Nice, that's a handy trick I didn't know!

bjorng commented:
I have taken a quick look and marked a few things in the code directly. Here follows some more general comments:

The comment characters are important for auto-indent to work in Emacs. Use one percent character (`%`) for comment at the end of a line of code. Use two percent characters (`%%`) for comments before a function or before a code block.

In patterns, put the alias variable to the right of the pattern like this:

```
case Expr of
  {ok, #file_info{}=Fi} ->...
end.

```

The indentation is is off in some places. Even if you don't use Emacs as your editor, please use it to fix up the indentation.

My opinions of the items above are **strong**.

Here are a few other things you may consider if you wish:

In guards, use `;` instead of `orelse` (at the top-level).

When having series of numbered variables, number them like this:

```
Pax0 = ...,
Pax1 = ...,
Pax2 = ...,

```
(That is, the first variable should not be named `Pax`.)

bitwalker commented:
@bjorng I do use Emacs, but my configuration may be off for some reason, but I agree with all your notes, I'll get those fixed up and push an update ASAP.

bjorng commented:
Regarding the use of `ram_file`: We have been thinking about eliminating it in some future release. Therefore, we would not want new usages of it.

I think that you can use  a binary in exactly the same way as you use `ram_file`. Create a 512-byte binary; implement a write function that splits the block into three pieces using the bit syntax (the two to keep, and the middle one to discard) and combine the pieces to a new binary. It might not be slower than `ram_file`, because there is a certain overhead of calling a driver. You could also optimize the binary handling by starting out with an empty binary and expand it as needed.

bitwalker commented:
@bjorng I'll give that a shot. Could you link me to the discussion about removing ram_file? I'm surprised it would be removed, but I'm mostly curious about it since I didn't realize it was being deprecated. In any case, I think your recommendation should work just fine here, I was concerned it would be inefficient due to the amount of splitting and reconstructing of binaries, but if you think it's low overhead, I have no qualms about it.

bjorng commented:
So far, we have only discussed the removal internally in the Erlang/OTP team. `ram_file` has not been deprecated, so it will not be removed in the near future.

bitwalker commented:
@bjorng I've updated the PR with the following changes:

- Fixed indentation
- Fixed incorrect comment usage (`%` vs `%%` and vice versa)
- Moved all pattern aliases to right-hand side
- Switched `orelse` for `;` in function guards
- Did a pass to fix up numbered bindings to start with 0
- Rewrote the bit where I was using a ram file to just split/join a binary (though there is still legacy code for dealing with the `{binary, binary()}` argument to `add` which uses `ram_file`, I did not change that).
- Added more tests in general, and made some changes in a few places to fix bugs found
- Cleaned up some redundant code for sparse file handling

I've also re-run `tar_SUITE` and ran code coverage on the tests. We're up to 77% now, and many of those uncovered lines are clauses which return errors back to the caller, or clauses for functions like `month/1` of which we only hit one, but would behave the same regardless of which month it is. There is still some room for more tests, but I think the bulk of the important stuff is covered now.

Let me know if there are further changes you'd like me to make!

bjorng commented:
Thanks! We will test the branch in our daily builds. I'll have a closer look at the code next week, and I'll come back to you if there is anything more that should be changed.

bjorng commented:
We found a few failed test cases in our daily builds:

`xref_SUITE:init_per_SUITE/1` in the application `tools`. 

Multiple test cases in `ssh_sftp_SUITE` that use `erl_tar`.

(In general, we tried to avoid depending on other OTP applications to test features in some module, but it in this case it seems that there are features in `erl_tar` that is only tested in other applications. Sorry for that.)

Also, the `erl_tar.hrl` file is not included in a release. That causes a test case in `syntax_tools` to fail because it cannot find `erl_tar.hrl` when attempting to recompile `erl_tar.erl`. It can be corrected by changing a line in `lib/stdlib/Makefile` like this:

    INTERNAL_HRL_FILES= dets.hrl erl_tar.hrl

bjorng commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -225,809 +279,1581 @@ month(10) -> "Oct";
+
+is_ascii(Str) when is_list(Str) ->
+    not lists:any(fun (Char) -> Char >= 16#80 end, Str);
+is_ascii(<<>>) ->
```
This may seem a little pedantic.

Mixing binary matching with other clauses like this will cause non-optimal binary matching code. (See the Efficiency Guide.) If you break out the binary matching clauses to their own `is_ascii1/1` function the matching optimizations can be applied. (Use the `bin_opt_info` option to verify.)

That is:

```
is_ascii(Str) when is_list(Str) ->
    not lists:any(fun (Char) -> Char >= 16#80 end, Str);
is_ascii(Bin) ->
    is_ascii1(Bin).

is_ascii1(<<>>) ->
    true;
is_ascii1(<<C,_Rest/binary>>) when C >= 16#80 ->
    false;
is_ascii1(<<_, Rest/binary>>) ->
    is_ascii1(Rest).
```

bitwalker replied:
Good to know!

bjorng commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -225,809 +279,1581 @@ month(10) -> "Oct";
+
+foldl_read(TarName, Fun, Accu, #read_opts{}=Opts)
+  when is_function(Fun,4) ->
+    case catch open(TarName, [read|Opts#read_opts.open_mode]) of
```
Use `try`...`of` here instead `catch`.

Error exceptions should be passed through, not converted to `{error,Reason}`. That is, only throws should be catched.

bitwalker commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -14,198 +14,252 @@
+         extract/1, extract/2,
+         table/1, table/2, t/1, tt/1,
+         open/2, close/1,
+         parse_octal/1,
```
Ah, forgot to remove that during some testing, and it escaped my review, sorry about that!

bitwalker commented:
I'll make the requested changes and try to reproduce the test failures you mentioned, and update the PR when that's been done. Thanks for the feedback!

HansN commented:
Hi @bitwalker,

As background information, here are few words about how `ssh` is related to the `erl_tar` module:

There is a function [ssh_sftp:open_tar/3,4](http://erlang.org/doc/man/ssh_sftp.html#open_tar-3).  This function calls `erl_tar:init` with installs a fun that makes erl_tar call ssh_sftp low-level functions instead of the default file functions. The purpose is to enable a system with small primary and secondary memory to build a (huge) encrypted tar file on a remote sftp server **without building a tar-file locally**. There is also an requirement to be able to send chunks of a tar file member without first collecting the full member locally.  

Unfortunately there are no  test cases for this erl_tar:init in lib/stdlib/test/tar_SUITE.erl as bjorng remarked.

Reference: the ssh_ftp User's Guide [ SFTP Client with TAR Compression and Encryption](http://erlang.org/doc/apps/ssh/using_ssh.html#id62522).

bitwalker commented:
@bjorng I've updated the PR with your notes, and fixed the tests as best I can tell - locally the only test group which runs for `ssh_sftp_SUITE` (and tests `erl_tar`) is `remote_tar` - but there are other groups which test the same functionality (or appear to), but are skipped - it's not clear to me how to get those to run, but since all of the `remote_tar` tests are passing now, I'll have to rely on you to tell me what I'm missing to ensure the rest of those tests are good to go.

bjorng commented:
Thanks! I'll this PR to be tested in our daily builds again.

bjorng commented:
The test cases now pass. I'll try to have a second look at the code before the end of this week.

margnus1 commented on lib/stdlib/src/Makefile:
```diff
@@ -130,7 +130,7 @@ HRL_FILES= \
 
-INTERNAL_HRL_FILES= dets.hrl
+INTERNAL_HRL_FILES= dets.hrl erl_tar.hrl
 
```
You should add a dependency from `erl_tar.beam` to `erl_tar.hrl` at the bottom of this makefile.

bjorng commented on lib/sasl/src/systools_make.erl:
```diff
@@ -1904,8 +1904,10 @@ del_tar(Tar, TarName) ->
 
 add_to_tar(Tar, FromFile, ToFile) ->
-    case erl_tar:add(Tar, FromFile, ToFile, [compressed, dereference]) of
+    case catch erl_tar:add(Tar, FromFile, ToFile, [compressed, dereference]) of
```
Do you have a good the reason for the `catch` here?

Seems that it is unnecessary. If it necessary, we prefer `try` in new case.

@sirihansen Any thoughts?

bitwalker replied:
There was a good reason, but I'm having trouble recalling specifically what it was - I think it may have been dialyzer related, I'll play around with removing it and see if I can reproduce the reason I added it. If it does need to be kept, I'll update it to use `try`

bjorng commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,383 @@
+%% Options used when adding files to a tar archive.
```
There should be a copyright header at the beginning of the file.

bitwalker replied:
Good call, forgot about that, I'll add it

bjorng commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,383 @@
+-define(TYPE_GNU_LONGLINK, $K).    % next file symlinks to a file with a long name
+-define(TYPE_GNU_SPARSE, $S).      % sparse file
+
+% Mode constants from tar spec
```
Should be: "%%". There are several others below, too.

bjorng commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,383 @@
+                         header_ustar().
+
+%% The overall tar reader, it holds the low-level file handle,
+%% it's access, position, and the I/O primitives wrapper.
```
"its"

bjorng commented:
There does not seem to be any way cause xattrs to be written. So it seems that the code for writing them can be removed.

It also seems that xattrs can be read and collected, but they are never actually used. So it seems that xattrs could just be discarded and the field for holding xattrs in `#tar_header{}` could be removed.

bjorng commented:
While looking at the code, I decided to extend the test suite.

So far I have pushed three commits that you can squash into your commit. First is a smoke test of the V7 format to cover some more code. Then I have added a test to make sure that a created has the expected format: plain USTAR or PAX. The last commit fixes some comment characters.

bitwalker commented:
@bjorng We could remove xattrs, since you are right, they aren't currently used because they are implementation specific attributes, and we get the things we care about from the PAX-standard extended attributes anyway (for which we don't use xattrs, they are just handled directly). I'll make that change.

bitwalker commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,383 @@
+          devminor = 0 :: non_neg_integer(),    % minor number of character or block device
+          atime :: calendar:datetime(),         % access time
+          ctime :: calendar:datetime(),         % status change time
+          xattrs = #{} :: map()                 % extended attributes
```
Good catch, I'll update it.

bitwalker commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,383 @@
+                               {ok, string() | binary()} | eof | {error, term()}).
+-type position_fun() :: fun((position, {file:io_device() | term, non_neg_integer()}) ->
+                                   {ok, non_neg_integer()} | {error, term()}).
+-type file_op() :: write_fun()
```
Ah interesting, ok, I'll think on this one a bit and update it. I'm not sure if specifying `file_op()` as you've described is useful as a typespec, since it's unlikely to catch the mistakes I was attempting to catch.

bitwalker commented on lib/stdlib/src/erl_tar.erl:
```diff
@@ -225,809 +280,1613 @@ month(10) -> "Oct";
+                                     Devmin, ?PAX_NONE, Pax9),
+    {Block13, Pax11} = set_path(Block12, Pax10),
+    Pax12 = lists:foldl(fun ({K,V}, Acc) ->
+                                maps:put(<<?PAX_XATTR_STR,K>>, V, Acc)
```
Great catch! As it turns out, we'll never be writing any xattrs which are not PAX-standard, so this code can be removed anyway.

bitwalker commented:
@bjorng I've updated the PR with the changes requested, let me know if anything further is desired!

bjorng commented on lib/stdlib/src/erl_tar.hrl:
```diff
@@ -0,0 +1,394 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 1997-2017. All Rights Reserved.
```
Remove "1997-".

bitwalker commented:
@bjorng Done!

bjorng commented on lib/stdlib/doc/src/erl_tar.xml:
```diff
@@ -104,21 +107,20 @@
+        length of 100 bytes. USTAR uses a header field (prefix) in addition to the name field, and
+        splits file paths longer than 100 bytes into two parts. This split is done on a directory boundary,
+        and is done in such a way to make the best use of the space available in those two fields, but in practice
+        this will often mean that you have less than 255 bytes for a path. <c>erl_tar</c> wil
```
Last word: "will"

bjorng commented:
Found a spelling error when taking a last quick look at the documentation.

When you have fixed that, I intend to merge this PR.

bitwalker commented:
I apparently *really* need a spellchecker enabled haha. I've updated the PR. Thanks!

slashdotdash commented:
Did this change make it into release 19.3 as I cannot see mention of 
OTP-14226 in the [release README](http://erlang.org/download/otp_src_19.3.readme)?

bjorng commented:
@slashdotdash No, it will be included in 20.0.

ericmj commented:
It seems like with this change OTP20 can create tarballs that we cannot open in OTP19. This is obviously a huge problem for a package manager such as Hex.

Is there a flag for a compatibility mode I can set? If not what checks should I do on the input files to erl_tar to ensure we create an archive that can be opened in OTP19?

bitwalker commented:
@ericmj Unless things have changed since my PR, `erl_tar` should default to generating USTAR tarballs unless a newer format (i.e. PAX) is required due to filename length (though I suppose it's possible other metadata such as timestamps might cause this upgrade as well, but I don't believe that's the case). I just ran a test locally to double check that behaviour with the release tarballs generated by Distillery - a release tarball produced by 20.0 can be extracted by 19.3. So it's not happening *every* time - at least on my machine (macOS). In any case, you are correct that it's a problem regardless.

There is no compat option (at least not that I added in my original PR, we just try to remain compatible by default). The check to perform is whether the filename length (the last portion of the path) is greater than 100 characters, or whether the total path in the tarball would exceed 255 characters in length - this will require PAX headers when in OTP 20 will be automatically used in that case - to remain compatible with OTP 19, you'd need to raise an error in those cases.

ericmj commented:
Thanks for the explanation @bitwalker.

> I just ran a test locally to double check that behaviour with the release tarballs generated by Distillery - a release tarball produced by 20.0 can be extracted by 19.3. So it's not happening every time - at least on my machine (macOS). In any case, you are correct that it's a problem regardless.

I got the first report of the issue today, so it seems to be a rare occurrence at least.

> The check to perform is whether the filename length (the last portion of the path) is greater than 100 characters, or whether the total path in the tarball would exceed 255 characters in length - this will require PAX headers when in OTP 20 will be automatically used in that case - to remain compatible with OTP 19, you'd need to raise an error in those cases.

It looks like more checks are needed unfortunately, for example unicode in filenames, file size, uid/gid sizes and so on. The Hex package that broke for us seemed to be using PAX because the UID size was too big. Since the checks are so complicated we have decided to bundle erl_tar from OTP20 in Hex for Mix and it looks like Rebar3 will do the same.
